### DISCUSSION TITLE
Using `.extend` after doing `.superRefine`.

### DISCUSSION THREAD

**Discussion by cakoose** [colinhacks/zod#694] [Q&A]
You can `.extend()` from a `z.object(...)`, but not if you've added a refinement.

Example ([playground link](https://www.typescriptlang.org/play?ssl=14&ssc=4&pln=1&pc=1#code/JYWwDg9gTgLgBAKjgQwM5wF5wGZQiOAcgwgBNCBuAKCoGMIA7VeAV1QFMoBlWgC3ZDI4AXkwA6CACMAVu1owAFAG8qcOA2Qh2ALnHMowBgHMFASgA0quAOTAANroxj9hkxaoBfU85ZhOAJXZsQ3YFBQA3czh5AA9TEQA+OBU1YGw4CLEbezFDUnYYgHlsBUIAAUJ44Wq4AFoARniUtWiYGLFkUlIASVRUFlClenzHMQAtMl7+9gBhMnYxWjYYfCitPuQjHSIxXcIvajUPT1NqOkZmODZOAHVgGF5u0h5+QRErjm4+GyyYmHYGKRlFZgKRRi5jGZLAcgA)):

```ts
import * as z from 'zod';

const userSchema = z.object({
  name: z.string(),
  email: z.string(),
}).superRefine((v, ctx) => {
  if (v.email.indexOf('@') === -1) {
    ctx.addIssue({code: z.ZodIssueCode.custom, message: '...'});
  }
});

const userWithIdSchema = userSchema.extend({
  id: z.string(),
});
```

Is there another way to achieve this?  Or maybe it's possible to update Zod to allow this?

**Comment by cakoose** [comment]
If I split things out, I can get it to work.  It's just a little more repetitive than ideal, i.e. easier to mess things up ([playground link](https://www.typescriptlang.org/play?ssl=19&ssc=33&pln=1&pc=1#code/JYWwDg9gTgLgBAKjgQwM5wF5wGZQiOAcgwgBNCBuAKCpgE8wBTOAVVUajgF5MA6YAHbYOAHnpMI2OAFd2UAMoBjABaMQyAOrAYyiNJgBhVYoDWqAHzVFEAaniyOS1eq069h42e58IAIwBWjIowABQA3lRwcALIIIwAXHx2UIIA5iEAlAA0kXBqyMAANokYvMlpmTkAvhnUVNa29nJO+UZBJt4hAG6JbBxZcMEAHiW8AEqM2IJqjAKGMEMZ3OZwEVHAUt28+UX8AqSMQwDy2CGEAAKES1w3cAC0AIxLa1GDC7zIpKQAkqio0oxwtYDqMAFpkX7-RgGMiMXiKWQwfADOJ-ZCpBJEXjYwg1ahRKpUKp1Bp2GTNFT5bwOBSUlzaXT6NqmVBlaRMKATKYCQE0lrqZkmWq5eo2Mk01zKb6kfnIakU5yaBnuQWsw4wWakcK5YCkUblATpbJEjJsjlc6YhPl05CC2pAA)):

```ts
import * as z from 'zod';

type User = z.infer<typeof userSchemaWithoutChecks>;
const userSchemaWithoutChecks = z.object({
  name: z.string(),
  email: z.string(),
});

const userSchemaCheck = (v: User, ctx: z.RefinementCtx) => {
  if (v.email.indexOf('@') === -1) {
    ctx.addIssue({code: z.ZodIssueCode.custom, message: '...'});
  }
};

const userSchema = userSchemaWithoutChecks.superRefine(userSchemaCheck);
  
const userWithIdSchema = userSchemaWithoutChecks.extend({
  id: z.string(),
}).superRefine(userSchemaCheck);
```

**Reply by scotttrinh** [reply]
I think it's helpful to think about how `ZodEffect` fits into the parsing structure: Refinements have an input type that is based on the output type of the schema it is refining. That means if you extend the schema after refinement, where does the refinement get triggered?

After the output of the base schema, but before the input of the extended schema? That means the extension, which _can_ modify it's inputs through `preprocess` and `transform` pretty dramatically can't guarantee that the output object still conforms to the refinement.

After the output of the extended schema? That means the input type of the refinement might be incorrect, again due to possible transformations or overrides in the extension.

When refactoring the `transform` concept for Zod@3, it was decided that the safest thing to do is to return a `ZodEffect` after transformation or refinements to avoid gotchas like this. Sharing refinements with an import is slightly less ergonomic, but it at least makes it explicit when and how the refinement is being applied, and it requires that the refinement function account for whatever possible input type you plan on using it with.

---

Separately, I realize your example is contrived, but just for clarity, I think if a refinement is just checking a single property on an object, it's better to refine _that_ property rather than refining the object. That would at least fix the issue in your example:

```ts
import * as z from 'zod';

type User = z.infer<typeof userSchema>;
const userSchema = z.object({
  name: z.string(),
  email: z.string().refine((str) => str.indexOf('@') === -1),
  // or: email: z.string().email(),
});


const userWithIdSchema = userSchema.extend({
  id: z.string(),
});
```

**Reply by cakoose** [reply]
Oh whoops, my example was misleading.  The `.superRefine` check I'm doing actually involves two fields.

> After the output of the base schema, but before the input of the extended schema? That means the extension, which can modify it's inputs through `preprocess` and `transform` pretty dramatically can't guarantee that the output object still conforms to the refinement.

That seems fine to me, but maybe I'm missing something?

The situation doesn't seem any different to what's possible now.  Using your code snippet as an example, if I add `preprocess` or `transform` to `userIdWithSchema`, that can modify things so that the final `email` value doesn't conform to its refinement, right?

**Reply by scotttrinh** [reply]
> The situation doesn't seem any different to what's possible now. Using your code snippet as an example, if I add preprocess or transform to userIdWithSchema, that can modify things so that the final email value doesn't conform to its refinement, right?

Right, but all of that requires that you have any refinements have the correct input type still. And your `transform` function will have the correct input type since you're defining the refinement on the property level. And if you tried to chain additional `ZodString` methods after `refine`, you'll find that they also don't work for the same reasons.

Maybe an example would be helpful to understand why it's not as simple when your apply effects to an object.

```ts
// In a world with `extend` possible after `refine`
const userSchema = z.object({
  email: z.string(),
}).refine((user) => user.email.indexOf("@") !== -1)); // user is { email: string }; here

const numberEmail = userSchema.extend({
  email: z.number(), // user is { email: number }; here
});
```

If we had the `refine` function run after the original schema, but before the second one, the refinement is pointless. If the `refine` function runs after the second schema, it gets the wrong input type and would fail with a runtime `indexOf is not a function`.

Imagine a slightly more complicated example: one that depends on the type of a property:

```ts
const userSchema = z.object({
  createdAt: z.date(),
  updatedAt: z.date(),
}).refine((user) => { // user is { createdAt: Date; updatedAt: Date; }
  return user.createdAt.getTime() <= user.updatedAt.getTime();
});

const notUpdated = userSchema.extend({
  updatedAt: z.null(), // user is { createdAt: Date; updatedAt: null }
});
```

How should we run the `refine` function in this case? Or with a transform

```ts
const withNumberDates = userSchema.transform((user) => ({
  ...user,
  createdAt: user.createdAt.getTime(),
  updatedAt: user.updatedAt.getTime(),
});
```

In both cases, the input type of the refine function is no longer correct, so the function would generate runtime errors. As it is now, we require that if you want to share the same refinement in all three cases, you do something like:

```ts
function checkDates(createdAt: Date | number, updatedAt: Date | number | null): boolean {
  const coercedCreatedAt = createdAt instanceof Date
    ? createdAt.getTime()
    : createdAt;
  const coercedUpdatedAt = updatedAt instanceof Date
    ? updatedAt.getTime()
    : updatedAt;
  if (coercedUpdatedAt === null) {
    return true;
  }
  return coercedCreatedAt <= coercedUpdatedAt;
}
```

Which requires that you do the runtime checking to avoid runtime errors. Now, maybe your extensions don't create any of these weird situations, but unfortunately, there is not an easy (any?) way to keep tracking of the possibility of breaking refinement rules or transformations which can be totally arbitrary. Also consider the other `ZodObject` methods like `partial`—that completely changes how a `refine` would need to be written.

**Comment by ljunio** [comment]
I'm running into the same issue here in this case. What would be the preferred solution in this case if we were to do a refinement that involves 2 fields? Would it be to create a separate schema without refinements and declare the refinements only where needed?

**Reply by scotttrinh** [reply]
You can always use `refine`/`superRefine` on an object schema, just note that you will not be able to use any object-specific methods (like `extend`, `partial`, etc) after doing so. This is due to needing to lock down the input type of `refine`. So if you have a complicated object hierarchy and want to share some refinements, it's best to make very flexible refinements and export those for use in places that you need to extend/modify your base schema(s), if that makes sense. Feel free to post an example of what you're trying to do and we can make some suggestions about how you might want to tackle them.

**Reply by ljunio** [reply]
Hi @scotttrinh, thanks for the reply.

I was hoping to do the following:

```ts
const shapeSchema = z.object({
    length: z.number;
    width: z.number;
    area: z.number;
}).refine(data => {
    return data.length > 0 && data.width > 0;
}, { message: "Cannot calculate area"});

const rectangleSchema = shapeSchema.extend({
    isRectangle: z.boolean()
});
```
This is just a basic example of what I'm trying to do, in reality, the base schema here is something that I'd also use in more schemas than just the `rectangleSchema`

Thanks

**Reply by scotttrinh** [reply]
There are a few strategies here, but here is one I would recommend: export very permissive refinements.

```ts
export const canCalculateArea = (data: { length?: number; width?: number }) => {
  return data.length !== undefined && data.width !== undefined
    ? data.length > 0 && data.width > 0
    : false; // Or `true` if you want to allow these schemas to pass
};
```

This allows it be added to the downstream schemas `rectangleSchema`, `squareSchema`, but also possibly `triangleSchema` which might not even have those properties, or `partial` versions of those schemas, or schemas that used `omit` to remove one or both of those properties.

But, you'll need to add the refinement in each place you create the extension rather than to the base `shapeSchema`. If you also want a version of `shapeSchema` that you can use that needs that refinement, you'll next to export both a "base" schema that can be extended and the version with the refinement which can just be `baseSchema.refine(canCalculateArea)` to avoid duplication.

**Reply by ljunio** [reply]
Hey @scotttrinh, thanks for the explanation. Just wanted to make sure I'm understanding this correctly.

So refinements should not be extended because it's very possible that the field that the refinement depends on for validation will change types (from a transform for example - like your email example above). Once that happens, there isn't any guarantee anymore that the refinement will still be applicable to the transformation. Is that correct?

So in order to circumvent this, your suggestion would be to just create a validator function and use it on the extended schemas when needed. I'm guessing this is because, at this point, it's pretty easy to guarantee that the extended schemas have locked-in input data for the refinement to work with. Is that correct?

**Reply by scotttrinh** [reply]
That is all correct! It's definitely very restrictive and can be stifling when trying to build up self-contained schemas that extend each other while inheriting things like refinements and transforms, but we chose to go the route of making it more difficult in order to avoid foot-guns, incorrect types, and throwing runtime errors. To that end, you can think of the philosophy of zod as parse then validate.

**Reply by ljunio** [reply]
Gotcha, thanks for going into detail on that.

**Comment by froxx93** [comment]
I had the exact same issue, but understand your explanation on why you guys decided to handle it the way you did.

Anyway I have an idea on how to prevent the so called "foot-gun"-effect:
Would it be possible to – instead of completely blocking manipulations of the ouput schema – make the parts touched by e.g. `.superRefine` read-only?
In my mind this would work something like this:
```typescript
const SchemaBase = z.object({
  name: z.string(),
  email: z.string(),
  active: z.boolean(),
});
const RefinedSchemaBase = SchemaBase.superRefine((v, ctx) => {
  if (v.email.indexOf('@') === -1) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: '...',
      touchedPath: ['email'],  // <- this would be an indicator on what path to set read-only (maybe place somewhere else)
    });
  }
});
const FullSchema = RefinedSchemaBase
  .omit({
    email: true,  // <- this would get marked invalid by TS
    active: true,
  })
  .extend({
    name: z.string().min(2),
    email: z.string().email(),  // <- this as well
    id: z.string.uuid(),
  })
```
That way it would not be possible to interfere with the `.superRefine`, but you could still add base-refinements to your base, and extend your schema on top of that in a non-interfering way to make your schema structure more intuitive.

Would anything like this be possible?

**Reply by p-runge** [reply]
Maybe the `touchedPath` could also be sent to `superRefine` as an additional "dependency list"-parameter similar to React's `useEffect`, so something like this:

```typescript
const RefinedSchemaBase = SchemaBase.superRefine((v, ctx) => {
  if (v.email.indexOf('@') === -1) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: '...',
    });
  }
}, ['email']); <- like this
```

Ideally that parameter would also limit the properties of `v`. So when I remove `'email'` from the list, `v.email` would be undefined.
